Add Wifi bootstrapping terminals to Retail Guide (#183)
uyuni-project/uyuni-docs: modules/retail/pages/retail-deploy-terminals-other.adoc

+
----
/etc/sysconfig/network/ifcfg-wlan0
  file.managed:
   - contents: |
        WIRELESS_ESSID=<wireless_network_name>
        WIRELESS_WPA_PSK=<wireless_network_password>
# regenerate initrd
  cmd.run:
  - name: 'mkinitrd'
----
. Apply the Salt state to the terminal:
+
----
salt <terminal_salt_name> state.apply update-terminal-credentials
----

[[NOTE]]
====
If you are using a separate network for the boot phase, the managed file might need to be renamed, or extended to [path]``/etc/sysconfig/network/initrd-ifcfg-wlan0``.
====

.Procedure: Changing Wireless Credentials for Seperate Boot Network
. Write a custom Salt state named ``saltboot_hook`` to provide custom ``/etc/sysconfig/network/ifcfg-wlan0`` to the system image during boot:
+
----
{% set root = salt['environ.get']('NEWROOT') %}
{{ root }}/etc/sysconfig/network/ifcfg-wlan0:
  file.managed:
    - source:
      - <salt://any/file>
    - require:
      - saltboot: saltboot_fstab
    - require_in:
      - saltboot: boot_system
----
The hook is applied after the system image is deployed.
. Create a different version for use during booting, called ``/etc/sysconfig/network/initrd-ifcfg-wlan0``.

> This file can be maintained by Salt on running terminals, to provide wireless credentials for use only during bootstrap and booting.

Reviewer: This is not related to the state above, but that generally either during image building or when adding branch specific password, this file `/etc/sysconfig/network/initrd-ifcfg-wlan0` can be used to specify password for boot phase.

Author: So, is the first one (``/etc/sysconfig/network/ifcfg-wlan0``) for normal operation, and the second one (``/etc/sysconfig/network/initrd-ifcfg-wlan0``) for boot? I feel like I have these mixed up.

Reviewer: Yes, you are correct. Keep in mind that this setup is optional, however. If there is no `/etc/sysconfig/network/initrd-ifcfg-wlan0`, then `/etc/sysconfig/network/ifcfg-wlan0` will be used for both boot and normal operation.

Reviewer: Particularly for this snipped, the `/etc/sysconfig/network/initrd-ifcfg-wlan0` for boot cannot be effectively managed by salt using `saltboot_hook`. Boot wireless network must be already running for salt to start, and `saltboot_hook` state cannot permanently modify `initrd`.

Author: Got it. Can you take another look and see if I'm closer now?

. Power on the POS terminal.
. Boot from the USB device to begin bootstrapping.

== Bootstrap Terminals over a Wireless Network

For terminals that cannot be connected directly to the physical network, you can bootstrap them over a wireless network.
Wireless networks do not support PXE booting, so you must perform the initial booting and initialization of the wireless connection on the terminal using a USB device.

For more information about using USB devices to boot, see xref:retail:retail-deploy-terminals-other.adoc[].

[[WARNING]]
====
Bootstrapping across a wireless network could expose a security risk if you are using encrypted OS images.
The boot ``initrd`` image and the partition that contains ``/etc/salt`` must be stored unencrypted on the terminal.
This allows {smr} to set up the wireless network.
If this breaches your security requirements, you will need to use a separate production network, with network credentials managed by Salt, so that credentials are not stored on the terminal unencrypted.
====

.Procedure: Preparing Wireless-Enabled Boot Image
. Ensure that the ``dracut-wireless`` package is included  in the image template.

> . Provide the wireless credentials using the [command]``WIRELESS_ESSID`` and [command]``WIRELESS_WPA_PSK`` kernel options.

Reviewer: ```suggestion
```

This line really shouldn't be here.

For the wifi enabled image the only mandatory part is that it must have `dracut-wireless` package included (line above this one)
Then optional part is the line below this, adding wifi credentials to the image itself using files.
This line about kernel options is not relevant to the image building, only when terminal is booting up way after image has been built.

. OPTIONAL: If you use the same wireless credentials for the entire organization, you can set them permanently by creating or editing the ``etc/sysconfig/network/ifcfg-wlan0`` file to the image template, with these details:
+
----
# ALLOW_UPDATE_FROM_INITRD
WIRELESS_ESSID=<wireless network name>
WIRELESS_WPA_PSK=<wireless network password>
----
If you want to use different credentials for bootstrapping to what is used during normal operation, you can remove the ``ALLOW_UPDATE_FROM_INITRD`` line.
. Build the image.
. Prepare a USB device using this image, and boot the terminal.
For more information about using USB devices to boot, see xref:retail:retail-deploy-terminals-other.adoc[].

Alternatively, you can create the file and add it to the ``initrd``:
+
----
echo ./etc/sysconfig/network/ifcfg-wlan0 | cpio -H newc -o | gzip >> /srv/saltboot/boot/initrd.gz

> ----

Reviewer: The thing is that this procedure is done on branch servers. I think it should be visibly separated from the instructions how to build wifi enabled image.

. In the [path]``/srv/salt/`` directory, create a salt state called ``update-terminal-credentials.sls``, and copy the details from [path]``/etc/sysconfig/network/ifcfg-wlan0``.
At a command prompt on the terminal:
+
----
/etc/sysconfig/network/ifcfg-wlan0
  file.managed:
   - contents: |
        WIRELESS_ESSID=<wireless_network_name>
        WIRELESS_WPA_PSK=<wireless_network_password>
# regenerate initrd
  cmd.run:
  - name: 'mkinitrd'
----
. Apply the Salt state to the terminal:
+
----
salt <terminal_salt_name> state.apply update-terminal-credentials
----

[[NOTE]]
====
If you are using a separate network for the boot phase, the managed file might need to be renamed, or extended to [path]``/etc/sysconfig/network/initrd-ifcfg-wlan0``.
====

You can also instruct terminals to use a different set of wireless credentials during the boot process, to what they use during normal operation.
This is done by using custom Salt states.

First of all, you need to set the wireless details for normal operation.
This will become the default settings.
Then you can specify a second Salt state with the wireless details for use during the boot procedure.

> .Procedure: Using Different Wireless Credentials for Booting

Reviewer: ```suggestion
.Procedure: Using Different Wireless Credentials for Production Network
```

> . Write a custom Salt state named ``/etc/sysconfig/network/ifcfg-wlan0`` containing the wireless details for normal operation.
> This Salt state is managed by Salt using ``saltboot_hook``, which is applied after the system image is deployed.
> If no other wireless credentials are supplied, these details will be used by default:

Reviewer: ```suggestion
. Write a custom Salt state named ``/srv/salt/saltboot_hook.sls`` containing the wireless details for normal operation.
This Salt state is applied by Saltboot after the system image is deployed.
```

+
----
{% set root = salt['environ.get']('NEWROOT') %}
{{ root }}/etc/sysconfig/network/ifcfg-wlan0:
  file.managed:

>     - source:
>       - <salt://any/file>

Reviewer: ```suggestion
   - contents: |
        WIRELESS_ESSID=<wireless_network_name>
        WIRELESS_WPA_PSK=<wireless_network_password>
```

This will be better then vague source file.

    - require:
      - saltboot: saltboot_fstab
    - require_in:
      - saltboot: boot_system
----

> . Create a second Salt state named ``/etc/sysconfig/network/initrd-ifcfg-wlan0`` containing the wireless details to use during boot.
> This file can be maintained by Salt on running terminals, to provide wireless credentials for use only during bootstrap and booting.

Reviewer: ```suggestion
```

This doesn't work like that. It is not possible to control wireless network credentials for boot network this way.
Boot network credentials can be controlled only by ways mentioned in procedures above:
1) provide default credentials with the image during image build
2) append credentials directly to the initrd using cpio
3) pass credentials on the kernel command line

This paragraph is only about how to securely specify credentials to wireless network used during normal operations (I also call it `Production Network`, not sure if it is correct)

====

== Deploy Terminals with a Removable USB Device

If you do not want to boot terminals from the network, you can create a live USB image and deploy terminals using a removable USB storage device.
This is useful if you cannot boot your terminals from the network, or if you do not have a local {smr} branch server providing network services.

You can prepare a bootable USB device with the image and tools required to deploy a POS terminal using a remote {smr} branch server.
POS devices booted using the USB device are assigned to the {smr} branch server that created the USB device.

You can create the bootable USB device on the branch server directly, or on the {smr} Server.

.Procedure: Creating a Bootable USB Device
. On the {smr} branch server, at the command prompt, as root, create the POS image.
You need to specify the size of the image, in megabytes.
Ensure you allow at least 300{nbsp}MB:
+
----
salt-call image_sync_usb.create <usb image name> <size in MB>
----
. Insert the USB device into the {smr} branch server machine, and copy the image to the new location:
+
----

> dd if=<usb image name> of=<path to usb device>

Reviewer 2: to speed up, add bs=16K or something similar.

Reviewer: True, in the code we use `bs=1M`.

----

When you have the image on the USB drive, check that the terminals you want to deploy allow local booting.
You can check this by editing the Saltboot formula in the {smr} {webui}.
For more information about the Saltboot formula, see xref:salt:formula-saltboot.adoc[].

.Procedure: Deploying Images to the Terminals using USB
. Insert the USB device into the terminal.
. Power on the POS terminal.
. Boot from the USB device to begin bootstrapping.

== Bootstrap Terminals over a Wireless Network

For terminals that cannot be connected directly to the physical network, you can bootstrap them over a wireless network.
Wireless networks do not support PXE booting, so you must perform the initial booting and initialization of the wireless connection on the terminal using a USB device.

For more information about using USB devices to boot, see xref:retail:retail-deploy-terminals-other.adoc[].

> [[WARNING]]

Reviewer 2: ```suggestion
[WARNING]
```

. Append these commands to the kernel boot parameters:
+
----
WIRELESS_ESSID=<wireless_network_name>
WIRELESS_WPA_PSK=<wireless_network_password>
----

=== Change Wireless Credentials

After you have set the wireless credentials, you can change them as needed.
The way to do this is different if you use one company-wide network, or if you have each branch server on its own separate network.

.Procedure: Changing Wireless Credentials for Single Network
. Rebuild the boot image with updated credentials.
. Recreate the bootable USB device based on the new boot image.
. Boot terminal from new USB device.

.Procedure: Changing Wireless Credentials for Multiple Networks
. In the [path]``/srv/salt/`` directory, create a salt state called ``update-terminal-credentials.sls``, and enter the new wireless network credentials:
+
----
/etc/sysconfig/network/ifcfg-wlan0
  file.managed:
   - contents: |
        WIRELESS_ESSID=<wireless_network_name>
        WIRELESS_WPA_PSK=<wireless_network_password>
# regenerate initrd
  cmd.run:
  - name: 'mkinitrd'
----
. Apply the Salt state to the terminal:
+
----
salt <terminal_salt_name> state.apply update-terminal-credentials
----

> [[NOTE]]

Reviewer 2: ```suggestion
[NOTE]
```

====
If you are using a separate network for the boot phase, the managed file might need to be renamed, or extended to [path]``/etc/sysconfig/network/initrd-ifcfg-wlan0``.
====

=== Use Multiple Wireless Networks

> You can instruct terminals to use a different set of wireless credentials during the boot process, to what they use during normal operation.

Reviewer 2: Instead of "instruct" I'd use configure or set up.

If you provide wireless credentials using ``initrd`` files, you can create two different files, one for use during boot called ``initrd-ifcfg-wlan0``, and the other for use during normal operation, called ``ifcfg-wlan0``.

Alternatively, you can use custom Salt states to manage wireless credentials with ``saltboot-hook``.

First of all, you need to set the wireless details for normal operation.
This will become the default settings.
Then you can specify a second Salt state with the wireless details for use during the boot procedure.

.Procedure: Using Different Wireless Credentials for Production Network
. Write a custom Salt state named ``/srv/salt/saltboot_hook.sls`` containing the wireless details for normal operation.
This Salt state is applied by Saltboot after the system image is deployed.
+
----
{% set root = salt['environ.get']('NEWROOT') %}
{{ root }}/etc/sysconfig/network/ifcfg-wlan0:
  file.managed:
   - contents: |
        WIRELESS_ESSID=<wireless_network_name>
        WIRELESS_WPA_PSK=<wireless_network_password>
    - require:
      - saltboot: saltboot_fstab
    - require_in:
      - saltboot: boot_system
----

> [[NOTE]]

Reviewer 2: ```suggestion
[NOTE]
```

----
. . Copy the file to ``initrd`` on the branch server:
+
----
echo ./etc/sysconfig/network/ifcfg-wlan0 | cpio -H newc -o | gzip >> /srv/saltboot/boot/initrd.gz
----
. Check that the terminals you want to deploy allow local booting.
You can check this by editing the Saltboot formula in the {smr} {webui}.
For more information about the Saltboot formula, see xref:salt:formula-saltboot.adoc[].

.Procedure: Providing Wireless Credentials During Terminal Boot
. Mount the USB device on the terminal, and boot from it.
. Append these commands to the kernel boot parameters:
+
----
WIRELESS_ESSID=<wireless_network_name>
WIRELESS_WPA_PSK=<wireless_network_password>
----

=== Change Wireless Credentials

After you have set the wireless credentials, you can change them as needed.
The way to do this is different if you use one company-wide network, or if you have each branch server on its own separate network.

.Procedure: Changing Wireless Credentials for Single Network
. Rebuild the boot image with updated credentials.
. Recreate the bootable USB device based on the new boot image.
. Boot terminal from new USB device.

.Procedure: Changing Wireless Credentials for Multiple Networks

> . In the [path]``/srv/salt/`` directory, create a salt state called ``update-terminal-credentials.sls``, and copy the details from [path]``/etc/sysconfig/network/ifcfg-wlan0``.
> At a command prompt on the terminal:

Reviewer: ```suggestion
. In the [path]``/srv/salt/`` directory, create a salt state called ``update-terminal-credentials.sls``, and enter the new wireless network credentials:
```

This part is done on SUMA server.
